server/src/main/java/org/opensearch/search/aggregations/bucket/filter/FiltersAggregatorFactory.java

```diff
@@ -93,19 +93,33 @@ public FiltersAggregatorFactory(
      * created if the aggregation collects documents reducing the overhead of
      * the aggregation in the case where no documents are collected.
      *
-     * Note that as aggregations are initialsed and executed in a serial manner,
-     * no concurrency considerations are necessary here.
+     * Note: With concurrent segment search use case, multiple aggregation collectors executing
+     * on different threads will try to fetch the weights. To handle the race condition there is
+     * a synchronization block
      */
     public Weight[] getWeights(SearchContext searchContext) {
-        if (weights == null) {
-            try {
-                IndexSearcher contextSearcher = searchContext.searcher();
-                weights = new Weight[filters.length];
-                for (int i = 0; i < filters.length; ++i) {
-                    this.weights[i] = contextSearcher.createWeight(contextSearcher.rewrite(filters[i]), ScoreMode.COMPLETE_NO_SCORES, 1);
+        if (weights != null) {
+            return weights;
+        }
+
+        // This will happen only for the first segment access in the slices. After that for other segments
+        // weights will be non-null and returned from above
+        synchronized (this) {
```

Comment:
Since the `Weight` instances behind the `weights` array are now shared between many threads, is it safe to use like that?

Reply:
Thank you for confirming